v0/destinations/fb_custom_audience/transform.js

```diff
@@ -187,114 +186,149 @@ function prepareDataField(userSchema, userUpdateList) {
-  if (isDefinedAndNotNull(bufferPayload.is_raw)) {
-    if (typeof bufferPayload.is_raw === "boolean") {
-      payload.is_raw = bufferPayload.is_raw;
-    } else {
+function preparePayload(
+  userUpdateList,
+  userSchema,
+  paramsPayload,
+  isHashRequired
+) {
+  const prepareFinalPayload = paramsPayload;
+  const { data_source } = paramsPayload;
+  if (Array.isArray(userSchema)) {
+    prepareFinalPayload.schema = userSchema;
+  } else {
+    prepareFinalPayload.schema = [userSchema];
+  }
+
+  prepareFinalPayload.data = prepareDataField(
+    userSchema,
+    userUpdateList,
+    isHashRequired
+  );
+  if (isDefinedAndNotNull(prepareFinalPayload.is_raw)) {
+    if (typeof prepareFinalPayload.is_raw !== "boolean") {
       logger.debug("is_raw field is not boolean, so dropping it");
+      delete prepareFinalPayload.is_raw;
     }
   }
   if (
-    isDefinedAndNotNull(bufferPayload.type) &&
-    isDefinedAndNotNull(bufferPayload.sub_type) &&
-    typeFields.includes(bufferPayload.type) &&
-    subTypeFields.includes(bufferPayload.sub_type)
+    !(
+      isDefinedAndNotNull(data_source.type) &&
+      isDefinedAndNotNull(data_source.sub_type) &&
+      typeFields.includes(data_source.type) &&
+      subTypeFields.includes(data_source.sub_type)
+    )
   ) {
-    payload.data_source = {
-      type: bufferPayload.type,
-      sub_type: bufferPayload.sub_type
-    };
+    delete prepareFinalPayload.data_source;
   }
-  return payload;
+  return prepareFinalPayload;
 }
 
-function prepareResponse(message, destination, category, userUpdateCategory) {
+function prepareResponse(
+  message,
+  destination,
+  category,
+  userUpdateCategory,
+  isHashRequired
+) {
   const { accessToken, userSchema } = destination.Config;
   const prepareParams = {};
-  let bufferPayload = {};
-  bufferPayload = constructPayload(message, MAPPING_CONFIG[category.name]);
+  let sessionPayload = {};
+  let paramsPayload = {};
+  let dataSsource = {};
+  sessionPayload = constructPayload(
+    message,
+    MAPPING_CONFIG[CONFIG_CATEGORIES.SESSION.name]
+  );
   if (
     checkSubsetOfArray(
-      Object.getOwnPropertyNames(bufferPayload),
+      Object.getOwnPropertyNames(sessionPayload),
       sessionBlockField
     )
   ) {
-    prepareParams.session = setupSessionBlock(bufferPayload);
+    prepareParams.session = sessionPayload;
   }
   prepareParams.access_token = accessToken;
-  if (checkSubsetOfArray(schemaFiels, userSchema)) {
-    prepareParams.payload = preparePayload(
-      message.properties[userUpdateCategory],
-      userSchema,
-      bufferPayload
-    );
-  } else
-    throw new CustomError(
-      "One or more of the schema fields are not supported",
-      400
-    );
+  paramsPayload = constructPayload(message, MAPPING_CONFIG[category.name]);
+  dataSsource = constructPayload(
+    message,
+    MAPPING_CONFIG[CONFIG_CATEGORIES.DATA_SOURCE.name]
+  );
+  paramsPayload.data_source = dataSsource;
```

Comment:
Also as mentioned below there are some validation based on `fields` of the `data_source` object lets keep those validations here and set it in `paramsPayload`  if those passes, otherwise we will not set it. 
That way we don't need to delete this property later